torch/distributed/_composable/fsdp/_fsdp_state.py

```diff
@@ -65,6 +74,25 @@ def _lazy_init(self) -> None:
                     state._is_root = False
                 self._all_states.append(state)
         self._init_fqns()
+        self._init_shared_state()
+
+    def _init_shared_state(self) -> None:
+        # Setting the all-gather/reduce-scatter streams to be higher priority
+        # can help avoid some issues where their copies in/out are delayed and
+        # block computation
+        high_priority = -1
+        self._default_stream = torch.cuda.current_stream()
+        self._all_gather_copy_in_stream = torch.cuda.Stream(priority=high_priority)
+        self._all_gather_stream = torch.cuda.Stream(priority=high_priority)
```

Comment:
Curious about the purpose of `_all_gather_stream`. Since collectives kernels always run on nccl pg's internal stream, `_all_gather_stream` would just be an indirection for stream synchronization? Also IIUC the priority specified here won't affect the comm kernels.

Reply:
> Also IIUC the priority specified here won't affect the comm kernels.

Correct! The priority here does not affect the comm. kernels.

> Curious about the purpose of _all_gather_stream. Since collectives kernels always run on nccl pg's internal stream, _all_gather_stream would just be an indirection for stream synchronization?

Also correct! The `_all_gather_stream` here is a trick to deal with the fact that NCCL PG collectives sync the NCCL stream with the current stream before running the collective.

This `_all_gather_stream` is used for so-called implicit prefetching where the CPU issue order is still the normal order: layer 0 all-gather -> layer 0 forward compute -> layer 1 all-gather -> layer 1 forward compute etc.

If we did not have a separate `_all_gather_stream`, then calling layer 1 all-gather would have the NCCL stream sync with the default (compute) stream and undesirably wait for layer 0 forward compute. The `_all_gather_stream` really just serves as a dummy stream so that we do not introduce that false dependency between layer i all-gather and layer i-1 forward compute.

Reply:
Got it. This makes sense.

Would it make sense to fold `_all_gather_copy_in_stream` and `_all_gather_stream` into a single stream, so we use fewer streams and simpify synchronization? Say we have a single stream `s`:
- `s` waits for the param group to be updated by the optimizer
- Allocate all-gather input buffer on `s`
- Perform copy-in on `s`
- Issue and wait all-gather from `s`

Would this work/does this make sense?

Reply:
Having both the AG copy-in stream and AG stream allows overlapping the _next_ all-gather copy-in with the _current_ all-gather, which can improve throughput in the forward pass. (In the backward pass, the all-gather copy-in can overlap with reduce-scatter.)

You might wonder, does overlapping the next all-gather copy-in with the current all-gather require using more memory? The answer is yes! However, I felt that it was a worthy tradeoff because:
- In the forward pass, including this overlap increases FSDP's peak memory usage from ~2x 'layer' size to 3x: current all-gather buffer (running all-gather), current copy-out parameters, next all-gather buffer (for copy-in).
- However, for the typical case without frozen parameters, the backward pass has FSDP peak memory of ~3x 'layer' size anyway! (e.g., prefetch next all-gather, current parameters, and current gradients)
- Even in the frozen parameter case, the ~1x 'layer' size is arguably in the noise of existing FSDP's `recordStream` spikes, and the real peak memory is generally not in forward. As such, I would lean toward just doing this overlapping always. If it is a really large concern, we could consider exposing it as an option to the user.

Reply:
Using one combined all-gather stream:
![Screenshot 2024-01-30 at 7 50 24 PM](https://github.com/pytorch/pytorch/assets/31054793/026ca06b-4f54-4059-bb36-7ba3ab04f36b)

Using separate all-gather copy-in and all-gather streams:
![Screenshot 2024-01-30 at 7 51 35 PM](https://github.com/pytorch/pytorch/assets/31054793/ff67c86b-3967-434d-8ae9-86c25f8ff4df)

Reply:
Thanks for the elaborate answer!